R-package/R/callbacks.R

```diff
@@ -1,563 +1,923 @@
-
-#' Callback closure for logging the evaluation history
+#' For \link{xgb.cv}, folds are a list with a structure as follows:\itemize{
+#' \item `dtrain`: The training data for the fold (as an `xgb.DMatrix` object).
+#' \item `bst`: Rhe `xgb.Booster` object for the fold.
+#' \item `watchlist`: A list with two DMatrices, with names `train` and `test`
+#' (`test` is the held-out data for the fold).
+#' \item `index`: The indices of the hold-out data for that fold (base-1 indexing),
+#' from which the `test` entry in the watchlist was obtained.
+#' }
 #'
-#' @details
-#' This callback function appends the current iteration evaluation results \code{bst_evaluation}
-#' available in the calling parent frame to the \code{evaluation_log} list in a calling frame.
+#' This object should \bold{not} be in-place modified in ways that conflict with the
+#' training (e.g. resetting the parameters for a training update in a way that resets
+#' the number of rounds to zero in order to overwrite rounds).
 #'
-#' The finalizer callback (called with \code{finalize = TURE} in the end) converts
-#' the \code{evaluation_log} list into a final data.table.
+#' Note that any R attributes that are assigned to the booster during the callback functions,
+#' will not be kept thereafter as the booster object variable is not re-assigned during
+#' training. It is however possible to set C-level attributes of the booster through
+#' \link{xgb.attr} or \link{xgb.attributes}, which should remain available for the rest
+#' of the iterations and after the training is done.
 #'
-#' The iteration evaluation result \code{bst_evaluation} must be a named numeric vector.
+#' For keeping variables across iterations, it's recommended to use `env` instead.
+#' \item data The data to which the model is being fit, as an `xgb.DMatrix` object.
 #'
-#' Note: in the column names of the final data.table, the dash '-' character is replaced with
-#' the underscore '_' in order to make the column names more like regular R identifiers.
+#' Note that, for \link{xgb.cv}, this will be the full data, while data for the specific
+#' folds can be found in the `model` object.
+#'
+#' \item watchlist The evaluation watchlist, as passed under argument `watchlist` to
+#' \link{xgb.train}.
+#'
+#' For \link{xgb.cv}, this will always be `NULL`.
+#'
+#' \item begin_iteration Index of the first boosting iteration that will be executed
+#' (base-1 indexing).
 #'
-#' Callback function expects the following values to be set in its calling frame:
-#' \code{evaluation_log},
-#' \code{bst_evaluation},
-#' \code{iteration}.
+#' This will typically be '1', but when using training continuation, depending on the
+#' parameters for updates, boosting rounds will be continued from where the previous
+#' model ended, in which case this will be larger than 1.
 #'
-#' @seealso
-#' \code{\link{callbacks}}
+#' \item end_iteration Index of the last boostign iteration that will be executed
+#' (base-1 indexing, inclusive of this end).
```

Comment:
Is this useful somehow other than pre-allocating a vector to store the evaluation history? These types of vectors/lists are quite cheap that don't need pre-allocation, and users who write callback usually know exactly the number of boosting rounds (they also write the training code).  I think it's not necessary to have an additional parameter with a condition that we (and the user) need to check whether it matches the number of boosting rounds.

Reply:
Yes, it is useful if e.g. one wants to:
* Create a custom step size calculator or otherwise modify some other parameter(s) accordingly.
* Target a certain number of snapshots during training instead of every N rounds or so.
* Create re-usable callbacks for xgboost in other packages.
* Create a different type of booster with an external optimizer like [this guy did here](https://github.com/catboost/catboost/issues/1994#issuecomment-1898816616).

Among others.

It's also useful when it comes to creating custom code to combine results of previous a callback with new results after an update (e.g. it would help to determine whether to append or replace).